File: SH/Mock2/Mock_Q3.py
from collections import defaultdict


class GNode:
    def __init__(self, id, color = "W", p = None):
        '''
        [color spec]
        W: not visited
        G: visited but its neighbors are not visited 
        B: visited and all the neighbors are visited 
        '''
        self.id = id
        self.color = color
        self.parent = p

    def __str__(self):
        return self.id
  
    def __hash__(self):
        return hash(self.id)
    
    def reset(self):
        self.id = id
        self.color = "W"
        self.parent = None

def countcomponent(n, graph):
    # adj_mat for 
    # graph = [[0 for _ in range(n)] for _ in range(n)]
    # graph = defaultdict(list)
    # for edge in edges:
    #     graph[edge[0]].append(edge[1])
    #     graph[edge[1]].append(edge[0])
    # print(graph)

    def dfs(node):
        if node.color == "W":
            node.color = "G"
            for neighbor in graph.get(node, []):
                dfs(neighbor)
            node.color = "B"

    # Perform DFS to detect cyclic dependencies
    n = 0
    for node in graph.keys():
        if node.color == "W":
            n += 1
        dfs(node)
        
    return n


if __name__ == "__main__":
    print("Mock_Q1.py")
    A, B, C, D, E, F = GNode('A'), GNode('B'), GNode('C'), GNode('D'), GNode('E'), GNode('F')
    
    print("\nCase 1")
    n = 5
    edges = [[0,1], [1, 2], [3, 4]]
    G = dict()
    G[A], G[B], G[D] = [B], [C], [E]
    # res = countcomponent(n, edges) 
    res = countcomponent(n, G) 
    ans = 2
    print(res, "\n  ==> check: ", res==ans)

    print("\nCase 2")
    n = 5
    edges = [[0,1], [1, 2], [2, 3], [3, 4]]
    res = countcomponent(n, edges)  
    ans = 1
    print(res, "\n  ==> check: ", res==ans)

File: SH/Mock2/Mock_Q3.cpp
#include <vector>
#include <iostream>
#include <string>
#include <bits/stdc++.h>
using namespace std;

class GNode {
    /*
    * colors spec
    * W: not visited
    * G: visited but its neighbors are not visited 
    * B: visited and all the neighbors are visited
    */
public:
    int id;
    string color;
    vector<GNode*> neighbors;
    GNode* parent;

    GNode(int id, string color = "W", GNode* parent = nullptr) : id(id), color(color), parent(parent) {}

    void addNeighbor(GNode* neighbor) {
        neighbors.push_back(neighbor);
    }
};


void dfs(GNode* node, GNode* parent = nullptr) {
    node->color = "G"; // 방문 시작 (Gray)
    node->parent = parent; // 부모 노드 설정

    for (size_t i = 0; i < node->neighbors.size(); ++i) {
        GNode* neighbor = node->neighbors[i];
        if (neighbor->color == "W") { // 방문하지 않은 이웃
            dfs(neighbor, node); // 현재 노드를 부모로 설정하여 dfs 호출
        }
    }    
    node->color = "B"; // 방문 완료 (Black)
}

int countComponents(int n, vector<vector<int>>& edges){
    vector<GNode*> adj_list;

    // Initialize Node
    for (int i = 0; i < n; ++i) {
        adj_list.push_back(new GNode(i));
    }
    // Construct adjacency list
    // using auto
    // for(auto edge:edges){
    //     auto edge1 = edge[0];
    //     auto edge2 = edge[1];
    //     adj_list[edge2].push_back(edge1);
    // }
    int n_edge = edges.size();
    // Simple
    for (int i = 0; i<n_edge; i++){
        adj_list[edges[i][0]]->addNeighbor(adj_list[edges[i][1]]);
        adj_list[edges[i][1]]->addNeighbor(adj_list[edges[i][0]]);
    }

    int count = 0;
    // Should go over all the adj_list
    for(int i=0; i<adj_list.size(); i++){
        if(adj_list[i]->color == "W"){
            dfs(adj_list[i]);
            count++;
        }
    }
    
    // Clear Memory
    for (size_t i = 0; i < adj_list.size(); ++i) {
        delete adj_list[i];
    }

    return count;
}



int main() {
    printf("[CPP] 323. Number of Connected Components in an Undirected Graph\n");

    // using printf
    printf("\nCase 1");
    {
        int n = 5;
        vector<vector<int>> edges = {{0, 1}, {1, 2}, {3, 4}};
        int ans = 2;
        int res = countComponents(5, edges);
        printf("\nResult    : %d", res);
        printf("\nAnwser    : %d", ans);
        printf("\n-> CHECK  : %s", ans == res ? "PASSED" : "FAILED");
        printf("\n");
    }

    // using cout
    cout << "\nCase 2" << endl;
    {
        int n = 5;
        vector<vector<int>> edges = {{0, 1}, {1, 2}, {2, 3}, {3, 4}};
        int ans = 1;
        int res = countComponents(5, edges);
        cout << "Result    : " << res << endl;
        cout << "Anwser    : " << ans << endl;
        cout << "-> CHECK  : " << (ans == res ? "PASSED" : "FAILED") << endl;
    }
    return 0;
}